### PR TITLE
Dockerizing the loader

### PR DESCRIPTION
**Pull request by wpears**
Allows the loader to be run via docker, which will likely be the preferred installation / execution method going forward.

Includes work to allow environment variables to be mocked (in normal operation process.env gets passed to functions rather than accessed directly).

Also, includes work to allow "unauthorized" requests via S3 if credentials don't exist as env variables or via ~/.aws/credentials. This allows testing of functions that deal with S3 to work without credentials if the test bucket is public for GetObject and ListBucket requests (which is has been set to). This allows docker setup to be a bit simpler (don't have to mount ~/.aws as a volume or pass env variables to docker run).

### CODE REVIEW

**Comment by wpears** [comment]
Closes #79, #78, #77, #59, and #55.

**Comment by hkeeler** [comment]
Just as a note, this is using [geodata/gdal](https://registry.hub.docker.com/u/geodata/gdal/) as the base Docker image.

File: README.md [hunk 1]
```diff
@@ -12,16 +12,29 @@ Data is transformed from these standard formats and loaded into Elasticsearch (t
...
   - Run `npm install` from the project root
+
+  - Alternatively, [install Docker](https://docs.docker.com/installation/#installation) and run the following:
+    - To build the image:
+      `docker build --rm --tag=hmda/grasshopper-loader .`
+    - To run the image:
+      `docker run -ti --rm hmda/grasshopper-loader`
+    - To run tests with dockerized Elasticsearch:
+      `docker run -d --rm --name es elasticsearch`
+      `docker run -ti --rm --link es:elasticsearch hmda/grasshoppper-loader`
```
Comment (hkeeler) [inline README.md]: Extra `p` in `grasshopper`.

**Comment by hkeeler** [comment]
I suggest the following changes to `Dockerfile`:

```diff
-CMD /bin/bash
+# Switch back to "nobody" used in geodata/gdal image
+USER nobody
+
+ENTRYPOINT ["./grasshopper-loader.js"]
+CMD ["--help"]
```

1. Its a no-no to run as `root` I've switched the user back to `nobody`, which is the user from the parent [geodata/gdal Dockerfile](https://registry.hub.docker.com/u/geodata/gdal/dockerfile/).  It seems to run OK this way, though I've only done simple testing.

1. I think we should consider image to be more of an executable binary, and less of an interactive env.  I've seen the above pattern mentioned a few times, using `ENTRYPOINT` and a `CMD` default to `--help` to let the user know how they can interact with the image.  And you can always override this default with `--entrypoint` arg.

These suggestions come out of a couple best practices docs:

* https://docs.docker.com/articles/dockerfile_best-practices/
* http://www.projectatomic.io/docs/docker-image-author-guidance/

**Comment by hkeeler** [comment]
Do we really want to bake Docker-smarts into the loader?  I get that it makes for simpler default setup, but it seems like a strange coupling.

**Comment by hkeeler** [comment]
Review complete.  Holding off on merging until @wpears is back next week so we can chat about whether this is really the approach we want to take.

**Comment by wpears** [comment]
Running as nobody breaks if temp files need to be created (which is currently required for zip files.. though hopefully this soon won't be the case). So instead I make a user which gets ownership of the app files/directories.

Currently I think having the interactive bash prompt fits the loader's purpose (and gets you closer to error messages, etc. that the loader may spit out). I would be okay providing the loader options to the docker run command, but I think we should talk about it/test it a bit more.

**Comment by wpears** [comment]
Also, you can just tack on loader commands to docker run, eg:

`docker run -ti --rm --link es:elasticsearch hmda/grasshopper-loader ./grasshopper-loader.js -d test/data/arkansas.json`

I could see the default loader CMD being `./grasshopper-loader.js --help` without setting an ENTRYPOINT, which would allow users to get bash if they want to run repeated commands in the container (testing GDAL, etc.). I still kinda like giving users bash by default though, as I think it feels more like "okay, the environment has set up, now I can play around and see what's happened."

**Comment by hkeeler** [comment]
I wasn't really seeing the loader as a tool that users would shell into and perform multiple commands against.  If you think that is a common use case, which it sounds like it is, then yes the setup you've mentioned does make sense.  In the end it doesn't matter much since any of the Dockerfile's `CMD` and `ENTRYPOINT` setup can be overridden by `docker run`.

**Pull request merged**